### PR TITLE
[docs] Update notifications sample code

### PR DESCRIPTION
**Pull request by douglowder**
# Why

Improve notifications documentation sample code:

- Convert to TypeScript
- Better error handling and typechecking
- Better info comments

# Test Plan

Docs CI should pass

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: docs/pages/push-notifications/push-notifications-setup.mdx [hunk 1]
```diff
@@ -90,68 +93,92 @@ async function registerForPushNotificationsAsync() {
...
+  const responseListener = useRef<Notifications.Subscription>();
 
   useEffect(() => {
-    registerForPushNotificationsAsync().then(token => setExpoPushToken(token));
-
-    /* @info This listener is fired whenever a notification is received while the app is foregrounded. */
-    notificationListener.current = Notifications.addNotificationReceivedListener(notification => {
-      setNotification(notification);
-    });
+    /* @info Gets the push token and shows it in the UI, or shows the error if one occurs */
```
Comment (amandeepmittal) [inline docs/pages/push-notifications/push-notifications-setup.mdx:142]:
```suggestion
    /* @info Gets the push token and displays it in the UI, or in case of an error, displays the error message. */
```

File: docs/pages/push-notifications/push-notifications-setup.mdx [hunk 2]
```diff
@@ -40,7 +40,7 @@ Run the following command to install the `expo-notifications`, `expo-device` and
 
 The code below shows a working example of how to register for, send, and receive push notifications in a React Native app. Copy and paste it into your project:
 
-```jsx App.js
+```jsx App.tsx
```
[the remark this answers is not in the document]
Comment (douglowder) [reply docs/pages/push-notifications/push-notifications-setup.mdx:43]: Good catch

File: docs/pages/push-notifications/push-notifications-setup.mdx [hunk 3]
```diff
@@ -57,8 +57,8 @@ Notifications.setNotificationHandler({
 });
 /* @end */
 
-// Can use this function below or use Expo's Push Notification Tool from: https://expo.dev/notifications
-async function sendPushNotification(expoPushToken) {
+/* @info Sends a notification to Expo. Expo's push notification tool at https://expo.dev/notifications may also be used. */
```
[the remark this answers is not in the document]
Comment (christopherwalter) [reply docs/pages/push-notifications/push-notifications-setup.mdx]: "Sends a notification to Expo" feels a bit unclear, but maybe it's just me. I'd prefer to clarify that we're sending something to Expo, and then Expo is sending that along to a recipient device. 

Maybe I'm overthinking this, but when I first onboarded to Push Notifications I did the demo and I didn't fully grasp that clicking the big blue "Send Notification" button actually sent anything anywhere. It seemed all local, like I was triggering a cute little demo notification to my app, from my app, all within my device. (And thus it felt a bit gimmicky, tbh). 

Anyway, I just feel like we could clarify that we're not sending a notification to Expo, but that Expo is sending a notification to _us_ (or whatever device we specify). 

Feel free to disregard this, sorry I don't have a specific suggestion on the copy.

File: docs/pages/push-notifications/push-notifications-setup.mdx [hunk 4]
```diff
@@ -90,68 +93,92 @@ async function registerForPushNotificationsAsync() {
...
-      projectId: Constants.expoConfig.extra.eas.projectId,
-    });
+    const projectId =
+      Constants?.expoConfig?.extra?.eas?.projectId ??
+      Constants?.easConfig?.projectId;
+    if (!projectId) {
+      handleRegistrationError('Project ID not found');
+    }
+    try {
+    /* @info This provides the ExpoPushToken using the unique project ID. */
```
Comment (christopherwalter) [inline docs/pages/push-notifications/push-notifications-setup.mdx:116]: IMO this could be clearer by saying something like `This fetches the ExpoPushToken (if not previously fetched), which is unique to this device and projectID`.

File: docs/pages/push-notifications/push-notifications-setup.mdx [hunk 5]
```diff
@@ -90,68 +93,92 @@ async function registerForPushNotificationsAsync() {
...
+      const pushTokenString = (
+        await Notifications.getExpoPushTokenAsync({
+          projectId,
+        })
+      ).data;
     /* @end */
-    console.log(token);
+      console.log(pushTokenString);
+      return pushTokenString;
+    } catch (e: any) {
```
Comment (christopherwalter) [inline docs/pages/push-notifications/push-notifications-setup.mdx:125]: nit: can we avoid `any`?

**Review by christopherwalter — APPROVED** [review]
Left some comments but nothing blocking. Consider @amandeepmittal's suggestions before landing.

File: docs/pages/push-notifications/push-notifications-setup.mdx (hunk 3 above)
[the remark this answers is not in the document]
Comment (amandeepmittal) [reply docs/pages/push-notifications/push-notifications-setup.mdx]: This is good context @christopherwalter!

Maybe we can rephrase this and just say "Sends a notification to the device. Can also use Expo's push notification tool to test sending a notification: https://expo.dev/notifications." 

I don't have much context on how this works here but we can also rephrase it to "Sends a notification to the device from Expo." as per your explanation. This way, the comment here will be clear about what action the function in subject performs.

**Review by amandeepmittal — APPROVED** [review]

File: docs/pages/push-notifications/push-notifications-setup.mdx (hunk 4 above)
Comment (douglowder) [reply docs/pages/push-notifications/push-notifications-setup.mdx:116]:
```suggestion
    /* @info This fetches the ExpoPushToken (if not previously fetched), which is unique to this device and projectID. */
```

File: docs/pages/push-notifications/push-notifications-setup.mdx (hunk 5 above)
Comment (douglowder) [reply docs/pages/push-notifications/push-notifications-setup.mdx:125]: @christopherwalter I'll have a look to see if that is possible. The error isn't always a string as I recall.

File: docs/pages/push-notifications/push-notifications-setup.mdx [hunk 6]
```diff
@@ -90,68 +93,92 @@ async function registerForPushNotificationsAsync() {
...
+        Notifications.removeNotificationSubscription(responseListener.current);
     };
   }, []);
 
   return (
-    <View style={{ flex: 1, alignItems: 'center', justifyContent: 'space-around' }}>
+    <View
+      style={{ flex: 1, alignItems: 'center', justifyContent: 'space-around' }}
+    >
       <Text>Your expo push token: {expoPushToken}</Text>
```
[the remark this answers is not in the document]
Comment (douglowder) [reply docs/pages/push-notifications/push-notifications-setup.mdx:172]:
```suggestion
      <Text>Your Expo push token: {expoPushToken}</Text>
```

[the remark this answers is not in the document]
Comment (douglowder) [reply docs/pages/push-notifications/push-notifications-setup.mdx:172]: I've attempted to make these more consistent.

File: docs/pages/push-notifications/push-notifications-setup.mdx [hunk 7]
```diff
@@ -90,68 +93,92 @@ async function registerForPushNotificationsAsync() {
...
+        await Notifications.getExpoPushTokenAsync({
+          projectId,
+        })
+      ).data;
     /* @end */
-    console.log(token);
+      console.log(pushTokenString);
+      return pushTokenString;
+    } catch (e: any) {
+      handleRegistrationError(e);
```
Comment (douglowder) [inline docs/pages/push-notifications/push-notifications-setup.mdx:126]:
```suggestion
      handleRegistrationError(`${e}`);
```

**Review by amandeepmittal — APPROVED** [review]
Thanks for these updates!
